src/modules/tls/tls_mod.c

```diff
@@ -496,3 +527,59 @@ int mod_register(char *path, int *dlflags, void *p1, void *p2)
 
 	return 0;
 }
+
+
+
+#ifndef OPENSSL_NO_ENGINE
+/*
+ * initialize OpenSSL engine in child process
+ * PKCS#11 libraries are not guaranteed to be fork() safe
+ *
+ */
+static int engine_init()
+{
+	/*
+	 * Example of configuration for AWS CloudHSM
+	 * # kamailio.cfg
+	 * modparam("tls", "engine", "gem")
+	 * modparam("tls", "engine_config", "/usr/local/etc/kamailio/tls_gemalto.conf")
+	 * modparam("tls", "engine_algorithms", "ALL")
+	 *
+	 * # OpenSSL config file: /usr/local/etc/kamailio/tls_gemalto.conf
+	 * # https://www.openssl.org/docs/manmaster/man5/config.html
+	 * # /usr/lib64/openssl/engines/libgem.so is the Gemalto OpenSSL engine for CloudHSM/SafeNet Luna
+	 *
+	 * kamailio = kamailio_init
+	 * [kamailio_init]
+	 * engines = engine_section
+	 * [engine_section]
+	 * gem = gemalto_section
+	 * [ gemalto_section ]
+	 * # engines support ctrl strings; this is Gemalto specific
+	 * # ENGINE_INIT = <slot>:<appid>:<appid>:password=<slot-password>
+	 * ENGINE_INIT = 0:23:24:password=1233-ABCD-EFGH-XY32
+	 */
+	LM_INFO("With OpenSSL engine support\n");
+	if (strncmp(engine_settings.engine.s, "NONE", 4)) {
+		int err = 0;
+		ENGINE_load_builtin_engines();
+		OPENSSL_load_builtin_modules();
+		if (strncmp(engine_settings.engine_config.s, "NONE", 4)) {
+			err = CONF_modules_load_file(engine_settings.engine_config.s, "kamailio", 0);
+			if (!err) goto error;
+		}
+		engine = ENGINE_by_id(engine_settings.engine.s);
+		if (!engine) goto error;
+		err = ENGINE_init(engine);
+		if (!err) goto error;
+		if (strncmp(engine_settings.engine_algorithms.s, "NONE", 4)) {
+			err = ENGINE_set_default_string(engine, engine_settings.engine_algorithms.s);
+			if (!err) goto error;
+		}
+		LM_INFO("OpenSSL engine %*s initialized\n", engine_settings.engine.len, engine_settings.engine.s);
+	}
+	return 0;
+error:
+	return -1;
```

Comment:
It would be great to add some error logging here, to give some hints to the user that the engine, the engine config or the algorithm setting failed.